Behavior difference: const T extends string = string binds to default instead of inferring never

## Steps to reproduce

Repro repository: https://github.com/birkskyum/tsgo-const-default-repro

Inline, two dev dependencies (`typescript` and `@typescript/native-preview`):

**`package.json`**

```json
{
  "name": "tsgo-const-default-repro",
  "private": true,
  "type": "module",
  "devDependencies": {
    "typescript": "^6.0.2",
    "@typescript/native-preview": "7.0.0-dev.20260502.1"
  }
}
```

**`tsconfig.json`**

```json
{
  "compilerOptions": {
    "target": "ES2022",
    "module": "ESNext",
    "moduleResolution": "Bundler",
    "strict": true,
    "skipLibCheck": true,
    "noEmit": true,
    "lib": ["ES2022"]
  },
  "include": ["./repro.ts"]
}
```

**`repro.ts`**

```ts
type Paths<T> = unknown extends T ? string : string

declare function f<
  R extends { p: any },
  const T extends string = string,
>(opts: {
  from?: (T & Paths<R['p']>) | Paths<R['p']>
  via: string extends T ? 'wide' : 'narrow'
}): void

f({ from: 'x',          via: 'narrow' })   // T = 'x',   via 'narrow' (both OK)
f({ from: 'x' as 'x',   via: 'narrow' })   // T = 'x',   via 'narrow' (both OK)
f({ from: 'x' as never, via: 'narrow' })   // T should be `never`
```

Run:

```sh
pnpm install
./node_modules/.bin/tsc      # see "Behavior with typescript@6.0"
./node_modules/.bin/tsgo     # see "Behavior with tsgo"
```

## Behavior with `typescript@6.0`

`typescript@6.0.3`. Compiles cleanly, exit `0`, no diagnostics. `T` is bound to `never` for the third call, `string extends never` is `false`, so `via`'s expected type narrows to `'narrow'`. Accepted.

## Behavior with `tsgo`

`@typescript/native-preview@7.0.0-dev.20260502.1` (also reproduces on `7.0.0-dev.20260421.2`):

```
repro.ts(13,25): error TS2322: Type '"narrow"' is not assignable to type '"wide"'.
```

`T` is bound to the constraint default `string` instead of `never`, so `string extends string` is `true` and the conditional widens `via`'s expected type to `'wide'`. Both `T = never` and `T = string` produce a structurally type-correct call (the `from` parameter type collapses to `string` in either binding), but `tsc` picks the more specific `never` while `tsgo` short-circuits to the constraint default.

## Additional information

### Required ingredients

Each of the four is load-bearing. Drop any one and both compilers agree.

1. `const T extends string = string` (with the `string` default).
2. A sibling generic parameter `R` constrained to a structural type. Replacing `R['p']` with a fixed (non-generic) indexed access, for example `type X = { p: any }; ...Paths<X['p']>`, makes the bug disappear.
3. The prop holding the inferred value is typed `(T & C) | C` where `C` goes through an indexed access into the *generic* `R`. A bare `T` does not trigger.
4. `Paths<T>` is a *conditional* type that resolves to `string`, namely `unknown extends T ? string : string`. Replacing it with the bare alias `type Paths<T> = string` makes the bug disappear, even though both branches of the conditional return `string`.

### Where this surfaces in real code

Observed against [`@tanstack/react-router@1.169.x`](https://www.npmjs.com/package/@tanstack/react-router): calls like `<Link from={fullPath as never} {...props} />` in [`packages/react-router/src/route.tsx#L162-L165`](https://github.com/TanStack/router/blob/main/packages/react-router/src/route.tsx#L162-L165) emit `TS2741`. There, `from` flows through `ConstrainLiteral<T, RoutePaths<TRouter['routeTree']>>` (the `(T & C) | C` shape above) with a `MakeToRequired` conditional on the same `T`.

### Related

Same family as #2797, fixed by #2803 ("Fixed JSX context-sensitive children discrimination for generic signatures"). That fix was scoped to JSX-children context-sensitive discrimination. This case is a plain function call (no JSX, no children) and is still live three months after #2803 merged.

Just tested, and the issue persist

| command | exit | output |
|---|---|---|
| `tsc` | `0` | clean |
| `tsc --stableTypeOrdering` | `0` | clean |
| `tsgo` | `2` | `repro.ts(25,25): error TS2322: …` |
| `tsgo --stableTypeOrdering` | `2` | same error |

I also tried via `compilerOptions.stableTypeOrdering: true` in `tsconfig.json`. The divergence reproduces regardless, so it doesn't appear to be explained by 6.0 vs 7.0 type ordering.

--- a/internal/checker/inference.go
+++ b/internal/checker/inference.go
@@ -98,10 +98,16 @@ func (c *Checker) inferFromTypes(n *InferenceState, source *Type, target *Type) 
 		}
 		return
 	}
-	if target.flags&TypeFlagsUnion != 0 && source.flags&TypeFlagsNever == 0 {
+	if target.flags&TypeFlagsUnion != 0 {
+		var sourceTypes []*Type
+		if source.flags&TypeFlagsUnion != 0 {
+			sourceTypes = source.Types()
+		} else {
+			sourceTypes = []*Type{source}
+		}
 		// First, infer between identically matching source and target constituents and remove the
 		// matching types.
-		tempSources, tempTargets := c.inferFromMatchingTypes(n, source.Distributed(), target.Distributed(), (*Checker).isTypeOrBaseIdenticalTo)
+		tempSources, tempTargets := c.inferFromMatchingTypes(n, sourceTypes, target.Distributed(), (*Checker).isTypeOrBaseIdenticalTo)
 		// Next, infer between closely matching source and target constituents and remove
 		// the matching types. Types closely match when they are instantiations of the same
 		// object type or instantiations of the same type alias.
@@ -386,7 +392,12 @@ func (c *Checker) inferToMultipleTypes(n *InferenceState, source *Type, targets 
 	typeVariableCount := 0
 	if targetFlags&TypeFlagsUnion != 0 {
 		var nakedTypeVariable *Type
-		sources := source.Distributed()
+		var sources []*Type
+		if source.flags&TypeFlagsUnion != 0 {
+			sources = source.Types()
+		} else {
+			sources = []*Type{source}
+		}
 		matched := make([]bool, len(sources))
 		inferenceCircularity := false
 		// First infer to types that are not naked type variables. For each source type we

--- a/testdata/baselines/reference/compiler/inferenceWithNeverSource1.js
+++ b/testdata/baselines/reference/compiler/inferenceWithNeverSource1.js
@@ -1,0 +1,24 @@
+//// [tests/cases/compiler/inferenceWithNeverSource1.ts] ////
+
+//// [inferenceWithNeverSource1.ts]
+// https://github.com/microsoft/typescript-go/issues/3688
+type Paths<T> = unknown extends T ? string : string;
+
+declare function f<
+    R extends { p: any },
+    const T extends string = string,
+>(opts: {
+    from?: (T & Paths<R["p"]>) | Paths<R["p"]>;
+    via: string extends T ? "wide" : "narrow";
+}): void;
+
+f({ from: "x", via: "narrow" });
+f({ from: "x" as "x", via: "narrow" });
+f({ from: "x" as never, via: "narrow" });
+
+
+//// [inferenceWithNeverSource1.js]
+"use strict";
+f({ from: "x", via: "narrow" });
+f({ from: "x", via: "narrow" });
+f({ from: "x", via: "narrow" });

--- a/testdata/baselines/reference/compiler/inferenceWithNeverSource1.symbols
+++ b/testdata/baselines/reference/compiler/inferenceWithNeverSource1.symbols
@@ -1,0 +1,51 @@
+//// [tests/cases/compiler/inferenceWithNeverSource1.ts] ////
+
+=== inferenceWithNeverSource1.ts ===
+// https://github.com/microsoft/typescript-go/issues/3688
+type Paths<T> = unknown extends T ? string : string;
+>Paths : Symbol(Paths, Decl(inferenceWithNeverSource1.ts, 0, 0))
+>T : Symbol(T, Decl(inferenceWithNeverSource1.ts, 1, 11))
+>T : Symbol(T, Decl(inferenceWithNeverSource1.ts, 1, 11))
+
+declare function f<
+>f : Symbol(f, Decl(inferenceWithNeverSource1.ts, 1, 52))
+
+    R extends { p: any },
+>R : Symbol(R, Decl(inferenceWithNeverSource1.ts, 3, 19))
+>p : Symbol(p, Decl(inferenceWithNeverSource1.ts, 4, 15))
+
+    const T extends string = string,
+>T : Symbol(T, Decl(inferenceWithNeverSource1.ts, 4, 25))
+
+>(opts: {
+>opts : Symbol(opts, Decl(inferenceWithNeverSource1.ts, 6, 2))
+
+    from?: (T & Paths<R["p"]>) | Paths<R["p"]>;
+>from : Symbol(from, Decl(inferenceWithNeverSource1.ts, 6, 9))
+>T : Symbol(T, Decl(inferenceWithNeverSource1.ts, 4, 25))
+>Paths : Symbol(Paths, Decl(inferenceWithNeverSource1.ts, 0, 0))
+>R : Symbol(R, Decl(inferenceWithNeverSource1.ts, 3, 19))
+>Paths : Symbol(Paths, Decl(inferenceWithNeverSource1.ts, 0, 0))
+>R : Symbol(R, Decl(inferenceWithNeverSource1.ts, 3, 19))
+
+    via: string extends T ? "wide" : "narrow";
+>via : Symbol(via, Decl(inferenceWithNeverSource1.ts, 7, 47))
+>T : Symbol(T, Decl(inferenceWithNeverSource1.ts, 4, 25))
+
+}): void;
+
+f({ from: "x", via: "narrow" });
+>f : Symbol(f, Decl(inferenceWithNeverSource1.ts, 1, 52))
+>from : Symbol(from, Decl(inferenceWithNeverSource1.ts, 11, 3))
+>via : Symbol(via, Decl(inferenceWithNeverSource1.ts, 11, 14))
+
+f({ from: "x" as "x", via: "narrow" });
+>f : Symbol(f, Decl(inferenceWithNeverSource1.ts, 1, 52))
+>from : Symbol(from, Decl(inferenceWithNeverSource1.ts, 12, 3))
+>via : Symbol(via, Decl(inferenceWithNeverSource1.ts, 12, 21))
+
+f({ from: "x" as never, via: "narrow" });
+>f : Symbol(f, Decl(inferenceWithNeverSource1.ts, 1, 52))
+>from : Symbol(from, Decl(inferenceWithNeverSource1.ts, 13, 3))
+>via : Symbol(via, Decl(inferenceWithNeverSource1.ts, 13, 23))
+

--- a/testdata/baselines/reference/compiler/inferenceWithNeverSource1.types
+++ b/testdata/baselines/reference/compiler/inferenceWithNeverSource1.types
@@ -1,0 +1,54 @@
+//// [tests/cases/compiler/inferenceWithNeverSource1.ts] ////
+
+=== inferenceWithNeverSource1.ts ===
+// https://github.com/microsoft/typescript-go/issues/3688
+type Paths<T> = unknown extends T ? string : string;
+>Paths : Paths<T>
+
+declare function f<
+>f : <R extends { p: any; }, const T extends string = string>(opts: { from?: (T & Paths<R["p"]>) | Paths<R["p"]>; via: string extends T ? "wide" : "narrow"; }) => void
+
+    R extends { p: any },
+>p : any
+
+    const T extends string = string,
+>(opts: {
+>opts : { from?: (T & Paths<R["p"]>) | Paths<R["p"]>; via: string extends T ? "wide" : "narrow"; }
+
+    from?: (T & Paths<R["p"]>) | Paths<R["p"]>;
+>from : Paths<R["p"]> | (T & Paths<R["p"]>) | undefined
+
+    via: string extends T ? "wide" : "narrow";
+>via : string extends T ? "wide" : "narrow"
+
+}): void;
+
+f({ from: "x", via: "narrow" });
+>f({ from: "x", via: "narrow" }) : void
+>f : <R extends { p: any; }, const T extends string = string>(opts: { from?: (T & Paths<R["p"]>) | Paths<R["p"]>; via: string extends T ? "wide" : "narrow"; }) => void
+>{ from: "x", via: "narrow" } : { from: string; via: "narrow"; }
+>from : string
+>"x" : "x"
+>via : "narrow"
+>"narrow" : "narrow"
+
+f({ from: "x" as "x", via: "narrow" });
+>f({ from: "x" as "x", via: "narrow" }) : void
+>f : <R extends { p: any; }, const T extends string = string>(opts: { from?: (T & Paths<R["p"]>) | Paths<R["p"]>; via: string extends T ? "wide" : "narrow"; }) => void
+>{ from: "x" as "x", via: "narrow" } : { from: "x"; via: "narrow"; }
+>from : "x"
+>"x" as "x" : "x"
+>"x" : "x"
+>via : "narrow"
+>"narrow" : "narrow"
+
+f({ from: "x" as never, via: "narrow" });
+>f({ from: "x" as never, via: "narrow" }) : void
+>f : <R extends { p: any; }, const T extends string = string>(opts: { from?: (T & Paths<R["p"]>) | Paths<R["p"]>; via: string extends T ? "wide" : "narrow"; }) => void
+>{ from: "x" as never, via: "narrow" } : { from: never; via: "narrow"; }
+>from : never
+>"x" as never : never
+>"x" : "x"
+>via : "narrow"
+>"narrow" : "narrow"
+

--- a/testdata/tests/cases/compiler/inferenceWithNeverSource1.ts
+++ b/testdata/tests/cases/compiler/inferenceWithNeverSource1.ts
@@ -1,0 +1,16 @@
+// @strict: true
+
+// https://github.com/microsoft/typescript-go/issues/3688
+type Paths<T> = unknown extends T ? string : string;
+
+declare function f<
+    R extends { p: any },
+    const T extends string = string,
+>(opts: {
+    from?: (T & Paths<R["p"]>) | Paths<R["p"]>;
+    via: string extends T ? "wide" : "narrow";
+}): void;
+
+f({ from: "x", via: "narrow" });
+f({ from: "x" as "x", via: "narrow" });
+f({ from: "x" as never, via: "narrow" });

